platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/eventhandling/TransactionPrehandler.java

```diff
@@ -0,0 +1,107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.eventhandling;
+
+import static com.swirlds.logging.legacy.LogMarker.EXCEPTION;
+import static com.swirlds.metrics.api.Metrics.INTERNAL_CATEGORY;
+
+import com.swirlds.base.time.Time;
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.platform.event.GossipEvent;
+import com.swirlds.platform.internal.EventImpl;
+import com.swirlds.platform.state.nexus.SignedStateNexus;
+import com.swirlds.platform.state.signed.ReservedSignedState;
+import com.swirlds.platform.stats.AverageTimeStat;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.temporal.ChronoUnit;
+import java.util.Objects;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Performs the prehandling of transactions
+ */
+public class TransactionPrehandler {
+    private static final Logger logger = LogManager.getLogger(TransactionPrehandler.class);
+
+    /**
+     * A source to get the latest immutable state
+     */
+    private final SignedStateNexus latestImmutableStateNexus;
+
+    /**
+     * Average time spent in to prehandle each individual transaction (in microseconds)
+     */
+    private final AverageTimeStat preHandleTime;
+
+    private final Time time;
+
+    /**
+     * Constructs a new TransactionPrehandler
+     *
+     * @param platformContext           the platform context
+     * @param latestImmutableStateNexus the latest immutable state nexus
+     */
+    public TransactionPrehandler(
+            @NonNull final PlatformContext platformContext, @NonNull final SignedStateNexus latestImmutableStateNexus) {
+        this.time = platformContext.getTime();
+        this.latestImmutableStateNexus = Objects.requireNonNull(latestImmutableStateNexus);
+
+        preHandleTime = new AverageTimeStat(
+                platformContext.getMetrics(),
+                ChronoUnit.MICROS,
+                INTERNAL_CATEGORY,
+                "preHandleMicros",
+                "average time it takes to perform preHandle (in microseconds)");
+    }
+
+    /**
+     * Prehandles application transactions
+     *
+     * @param event the event to prehandle
+     */
+    public void prehandleApplicationTransactions(final GossipEvent event) {
+        // FUTURE WORK: As a temporary workaround, convert to EventImpl. This workaround will be removed as part of the
+        // event refactor
+        final EventImpl eventImpl = new EventImpl(event, null, null);
+
+        final long startTime = time.nanoTime();
+
+        ReservedSignedState latestImmutableState = null;
+        try {
+            latestImmutableState = latestImmutableStateNexus.getState("transaction prehandle");
+            while (latestImmutableState == null) {
+                latestImmutableState = latestImmutableStateNexus.getState("transaction prehandle");
+            }
```

Comment:
this has the potential to spin indefinitely and not provide any feedback. maybe after like 100 tries we should log something and sleep, then try again

Reply:
spinning indefinitely consumes a thread/cpu for useless work.  I presume this needs to be fixed before the PR can merge.

Reply:
I can improve this if there is consensus on correct strategy- but note that the spinny wait is just carrying forward the previous functionality.

Reply:
I lean towards this being outside the scope of this PR

Reply:
I've removed the block from the PR being merged.   

If the solution had stayed with the AtomicReference being updated by an asynchronous method call, I would have suggested using notify() and wait().  But with the introduction of the nexus,  what about a blocking get from the nexus?

Reply:
I think it would be a good future issue to add a blocking get from the nexus.

I created one here https://github.com/hashgraph/hedera-services/issues/11989